src/andotp_fmt_plug.c

```diff
@@ -0,0 +1,283 @@
+	{"$andotp$0*d5d82faa599a0aafde38918d*7b7f*3c244ae13d5b45198ed39bacc285f691", "åbc"},
+	{NULL}
+};
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int *saved_len;
+static int any_cracked, *cracked;
+static size_t cracked_size;
+
+static struct custom_salt {
+	uint32_t ctlen;
+	unsigned char iv[IVLEN];
+	unsigned char tag[TAGLEN];
+	unsigned char ciphertext[MAX_CIPHERTEXT_LENGTH];
+} *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	saved_len = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_len));
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	any_cracked = 0;
+	cracked = mem_calloc(cracked_size, 1);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(saved_len);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy, *keeptr, *p;
+	int value, extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+
+	ctcopy += TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "*")) == NULL) // version / type
+		goto err;
+	if (!isdec(p))
+		goto err;
+	value = atoi(p);
+	if (value != 0)
+		goto err;
+	if ((p = strtokm(NULL, "*")) == NULL)   // iv
+		goto err;
+	if (hexlenl(p, &extra) != IVLEN * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "*")) == NULL)   // ciphertext
+		goto err;
+	if (hexlenl(p, &extra) < MIN_CIPHERTEXT_LENGTH * 2 || extra)
+		goto err;
+	if (hexlenl(p, &extra) > MAX_CIPHERTEXT_LENGTH * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "*")) == NULL)   // tag
+		goto err;
+	if (hexlenl(p, &extra) != TAGLEN * 2 || extra)
+		goto err;
+
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	char *p;
+	int i;
+
+	memset(&cs, 0, SALT_SIZE);
+	ctcopy += TAG_LENGTH;
+	p = strtokm(ctcopy, "*");
+	p = strtokm(NULL, "*");
+	for (i = 0; i < IVLEN; i++)
+		cs.iv[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "*");
+	cs.ctlen = strlen(p) / 2;
+	for (i = 0; i < cs.ctlen; i++)
+		cs.ciphertext[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "*");
+	for (i = 0; i < TAGLEN; i++)
+		cs.tag[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+
+	MEM_FREE(keeptr);
+
+	return &cs;
+}
+
+static int check_password(int index, struct custom_salt *cs)
+{
+	unsigned char key[32];
+	unsigned char plain[MAX_CIPHERTEXT_LENGTH];
```

Comment:
@magnumripper 

This stack buffer is too large, and we don't even care about it contents!

I am going to apply the following patch,

``` diff
diff --git a/src/aes_gcm.h b/src/aes_gcm.h
index a57848a40..4ab8f1f2a 100644
--- a/src/aes_gcm.h
+++ b/src/aes_gcm.h
@@ -26,7 +26,8 @@ int aes_gcm_ae(const uint8_t *key, size_t key_len, const uint8_t *iv, size_t iv_
  */
 int aes_gcm_ad(const uint8_t *key, size_t key_len, const uint8_t *iv, size_t iv_len,
               const uint8_t *crypt, size_t crypt_len,
-              const uint8_t *aad, size_t aad_len, const uint8_t *tag, uint8_t *plain);
+              const uint8_t *aad, size_t aad_len, const uint8_t *tag, uint8_t *plain,
+              int skip_output);

 int aes_gmac(const uint8_t *key, size_t key_len, const uint8_t *iv, size_t iv_len,
             const uint8_t *aad, size_t aad_len, uint8_t *tag);
diff --git a/src/aes_gcm_plug.c b/src/aes_gcm_plug.c
index a1dc42f06..bb2c9b003 100644
--- a/src/aes_gcm_plug.c
+++ b/src/aes_gcm_plug.c
@@ -331,7 +331,8 @@ int aes_gcm_ae(const uint8_t *key, size_t key_len, const uint8_t *iv, size_t iv_
  */
 int aes_gcm_ad(const uint8_t *key, size_t key_len, const uint8_t *iv, size_t iv_len,
               const uint8_t *crypt, size_t crypt_len,
-              const uint8_t *aad, size_t aad_len, const uint8_t *tag, uint8_t *plain)
+              const uint8_t *aad, size_t aad_len, const uint8_t *tag, uint8_t *plain,
+              int skip_output)
 {
        uint8_t H[AES_BLOCK_SIZE];
        uint8_t J0[AES_BLOCK_SIZE];
@@ -343,7 +344,8 @@ int aes_gcm_ad(const uint8_t *key, size_t key_len, const uint8_t *iv, size_t iv_
        aes_gcm_prepare_j0(iv, iv_len, H, J0);

        /* P = GCTR_K(inc_32(J_0), C) */
-       aes_gcm_gctr(&ctx, J0, crypt, crypt_len, plain);
+       if (!skip_output)
+               aes_gcm_gctr(&ctx, J0, crypt, crypt_len, plain);

        aes_gcm_ghash(H, aad, aad_len, crypt, crypt_len, S);

diff --git a/src/andotp_fmt_plug.c b/src/andotp_fmt_plug.c
index c92651d47..42d825f66 100644
--- a/src/andotp_fmt_plug.c
+++ b/src/andotp_fmt_plug.c
@@ -168,7 +168,7 @@ static int check_password(int index, struct custom_salt *cs)
        SHA256_Update(&ctx, saved_key[index], saved_len[index]);
        SHA256_Final(key, &ctx);

-       ret = aes_gcm_ad(key, 32, cs->iv, IVLEN, cs->ciphertext, cs->ctlen, NULL, 0, cur_salt->tag, plain);
+       ret = aes_gcm_ad(key, 32, cs->iv, IVLEN, cs->ciphertext, cs->ctlen, NULL, 0, cur_salt->tag, plain, 1);

        if (!ret)
                return 1;
```

Reply:
This is done now. I also got a decent speed boost :)

Reply:
Well done sir!